modules/service/src/main/resources/db/migration/V0310__timing_windows.sql

```diff
@@ -0,0 +1,53 @@
+--- EXISTENCE
+CREATE TYPE e_timing_window_inclusion AS ENUM ('include', 'exclude');
+
+-- END AT/AFTER
+CREATE TYPE e_timing_window_end_type AS ENUM ('at', 'after');
+
+CREATE TABLE t_timing_window (
+  c_timing_window_id bigserial                 PRIMARY KEY,
+  c_observation_id   d_observation_id          NOT NULL,
+  c_inclusion        e_timing_window_inclusion NOT NULL,
+  c_start            timestamp                 NOT NULL,
+  c_end_at           timestamp,  -- if both c_end_at and c_end_after are null, remain open forever
+  c_end_after        interval,   --
+  c_repeat_period    interval,   -- if null, don't repeat
+  c_repeat_times     int4,       -- if null (and c_repeat_period not null), repeat forever
+
+  foreign key (c_observation_id) references t_observation(c_observation_id),
+
+  -- if window ends, either timestamp or duration is defined but not both.
+  constraint end_at_nand_end_after
+  check (num_nulls(c_end_at, c_end_after) >= 1),
```

Comment:
If it was `<=1`, then `num_nulls(...)` could be `0`, which means both are defined (which is what we want to avoid). We want to allow either 1 of them to be null (window ends), or both of them (window never ends).

Reply:
Ah right, thanks, I was reading it wrong.  :+1: